Add janis converter

illusional opened the pull request:
A few questions still to answer:

- Optional nested types, eg: `String?[]`
- Binding the prefix to each element of an array?
- Tool arguments (maybe this isn't applicable)

illusional committed:
Add initial JanisGenerator

illusional committed:
Add JanisGenerator import to make WrapperGenerator.__subclasses__ work

illusional committed:
Add janis as "pipe --format" option

illusional committed:
Add "validate_janis" method which runs the python file

illusional committed:
Add test_pipe_janis test

illusional committed:
Add janis-core (>= v0.11.2) as dependency

illusional committed:
Test reformatted files

multimeric commented:
> Optional nested types, eg: String?[] 

If this means "a list of optional strings", then no, aCLImatise doesn't support that. Can you think of an example CLI tool that has such a type? It might be a future direction to move the required/optional into the type object, since it's currently in the `Flag`/`Positional`

> Binding the prefix to each element of an array?

As above. Could be supported but would need an example

> Tool arguments (maybe this isn't applicable)

What do you mean by this?

illusional commented:
> Optional nested types, eg: String?[]

I looked through our tool definitions, I don't have an example for this.

> Binding the prefix to each element of an array?

Yep, **Gatk4ApplyBQSR**:

```bash
gatk4 ApplyBqsr -I bam -L interval1.bed -L interval2.bed
```

> Tool arguments

In Janis, these are:

1. "Unchangeable parameters" (eg: `--runMode alignReads`) for our StarAlignReads tool, or
2. "where you should perform some expression" (similar to CWL's `valueFrom`, we insist on doing this operation in an argument. But don't need these either.

multimeric commented:
Okay, both of those remaining points could conceivably be parsed from the help text, so I might eventually get around to them. But since I don't support that yet you don't have to worry about converting them for now.

illusional committed:
Add janisgenerator to __all__ export

illusional commented:
Hi @TMiguelT, it looks like the CI is failing with an `isort` error. It's trying to move the `deprecated` import, though when I ran `isort` locally this is where it placed it -> I'm running `VERSION 5.7.0` of `isort` if that helps. Not sure what to do because moving the import manually will cause my precommit hook to fail.

multimeric commented:
Oh yes this is a strange and annoying error. I've been chucking the third party libs in the `.isort.cfg` which seems to sort it.

illusional committed:
Manually move deprecated import to silence isort

illusional committed:
Fix Janis import

illusional commented:
Hey @TMiguelT, tests are passing now. Are there any other steps to follow?

multimeric commented:
Nope, just my (upcoming) code review, and it might be worth eyeballing the outputs to see if they make sense (and incorporating anything you note into the validation function)

multimeric commented on aclimatise/converter/janis.py:
```diff
@@ -0,0 +1,113 @@
+
+        return tool
+
+    def type_to_cwl_type(self, typ: cli_types.CliType, optional):
```
Is this function misnamed? Should it be janis type?

multimeric replied:
Also would appreciate a type annotation on the args and output here

illusional replied:
It should be `janis_type` and I've added type annotations.

multimeric commented on aclimatise/converter/janis.py:
```diff
@@ -0,0 +1,113 @@
+        return self.type_to_cwl_type(arg.get_type(), arg.optional)
+
+    def get_inputs(self, names: List[NamedArgument]):
+        import janis_core as j
```
Haven't you already imported janis at this point?

illusional replied:
I've removed this extra import

multimeric commented on aclimatise/converter/janis.py:
```diff
@@ -0,0 +1,113 @@
+            assert arg.name != "", arg
+            ret.append(
+                j.ToolInput(
+                    tag="in_" + arg.name,
```
You don't need to add `in_` if you namespace your inputs separately from your outputs. I only do this for CWL because all identifiers have to be globally unique

illusional replied:
We don't namespace inputs / outputs, so good to add this in here.

multimeric commented on setup.py:
```diff
@@ -21,6 +21,7 @@
         "psutil",
         "deprecated",
         "attrs",
+        "janis-pipelines.core >= 0.11.2, <= 0.12.0",
```
Out of interest, why the maximum version? If the API changes, should this not be using the latest API?

illusional replied:
I wasn't sure what you opinion of pinning to specific versions. Theoretically any version > = 0.11.2, I don't think we've changed the syntax of anything here in 18 of the past 24 months, so It's pretty safe. I've expanded the range.

multimeric replied:
I want to avoid pinning versions in case aCLImatise is itself used as a library and needs to not be too restrictive.

multimeric commented on test/util.py:
```diff
@@ -402,6 +402,20 @@ def validate_cwl(cwl: str, cmd: Command = None, explore: bool = True):
+        tmpfile.write_text(janis)
+
+        # just run the python file, and if it succeeds we're sweet
+        result = subprocess.run(["python", tmpfile])
```
I assume this is running the class definition. It's checking syntax which is good. Does it do any other validation?

illusional replied:
When converting to Janis, a translate is inserted at the bottom of the file, so running the python file will actually also translate it to WDL.

```python
if __name__ == "__main__":
    # or "cwl"
    YourTool().translate("wdl")
```

multimeric replied:
Sounds good!

multimeric reviewed (changes requested):
Looks good! Just a few comments. Incidentally I would appreciate type annotations wherever possible in the `JanisGenerator` class, cheers.

illusional committed:
Moved janis_core import to top + type annotations

I was hoping to avoid importing janis_core at the start to avoid any overhead. But less avoidable for type annotations.

illusional committed:
Open version of janis_core

illusional committed:
Format per GH action's request

illusional commented:
Thanks for the feedback @TMiguelT, I've addressed your feedback by importing janis_core at the top (rather than only when initialised), it provides the benefit of type annotations at the expense of some extra loading time of janis_core (but that shouldn't impact this).

multimeric commented:
At some point I'll work out a system that only conditionally imports the generator modules, meaning that users who don't want Janis won't have to install or load the Janis module. But you shouldn't have to worry about that. Also fwiw I believe you could have done the type annotations using strings to prevent the need for a global import.

Anyway, it's fine as it is.

illusional commented:
Tests all pass :)

multimeric commented:
Great!

multimeric commented:
Hey @illusional, I've had this merged into the Base Camp for a while, but I note that it's generating the Python class all on one line, e.g. https://aclimatise.github.io/BaseCamp//packages/samtools/1.11/samtools_dict/samtools_dict.py. Is there a simple fix for this?

illusional commented:
Thanks for question @TMiguelT! If the black formatted is installed, it will automatically format :)

multimeric commented:
As in I just have to passively have it installed when the Janis convert is run?

illusional commented:
Yep!
